hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/pipeline/Pipeline.java

```diff
@@ -140,6 +143,10 @@ public Instant getCreationTimestamp() {
     return creationTimestamp;
   }
 
+  public Instant getStateEnterTime() {
```

Comment:
IMO, need set closedState time when close is triggered for pipeline close, else this time is same as creation time as initialized in constructor.
This is to scrub from close time with delay.

Reply:
@sumitagrawl the {{PipelineState}} inside {{Pipeline}} class is immutable.
If we want to change the state of a pipeline, we create a new {{Pipeline}} object with new state which will also update the {{stateEnterTime}} with current time.

Since {{PipelineState}} is immutable, {{stateEnterTime}} is also designed to be immutable.

Reply:
@sumitagrawl the `PipelineState` inside `Pipeline` class is immutable.
If we want to change the state of a pipeline, we create a new `Pipeline` object with new state which will also update the `stateEnterTime` with current time.

Since `PipelineState` is immutable, `stateEnterTime` is also designed to be immutable.

Reply:
@nandakumar131 
We can add comment in Pipeline build() to notify the stateEnterTime will be latest when enter the state and no need update the time while building this object. Just to avoid later on changes if context is missed.

Reply:
Yeah, it is a little strange that creationTime and stateEnterTime are both basically hardcoded to "now". Feels like we should only need one of them, as creation time isn't really "correct". If the pipeline changes state from open to closed, the creation time is going to change too, due to its immutable nature. I can see how this could be confusing to someone in the future.

Reply:
@sodonnel, I completely understand the confusion here. We should refactor the code for better understandability.
Currently we don't have any functional issues as we update the creation time in the builder while we create the Pipeline object from existing pipeline reference.
If you look at `Pipeline$Builder(Pipeline pipeline)`, it assigns the creation time from the given pipeline reference.

With the current code, it's easy to introduce bugs in future. We should re-write the `Pipeline` class.

Reply:
The `creationTimestamp` in the `Pipeline` constructor is a fallback value. We update the creation time in the `Pipeline$Builder#build`. If the creation time is not set in the builder, we use the `Instant#now` value set in constructor.

Reply:
Ah, I missed that the builder sets creation time. I guess this is OK as the constructor is private, so the only way to create the object is via the builder.

I noticed that the setting is however public:

```
  public void setCreationTimestamp(Instant creationTimestamp) {
    this.creationTimestamp = creationTimestamp;
  }
```

We could make that private as the inner Builder class can call the private methods. This is not really in scope for this PR though.